### PR TITLE
Generate URL with predefined filters

### LINKED ISSUE #6739: Generate URL with predefined filters
| Wazuh | Rev  | Browser                      |
|-------|------|------------------------------|
| 4.9.0 | 00 | Chrome, Firefox, Safari, etc |

## Description
We need to develop a service that generates a valid URL with a set of preselected filters. Some of the use cases would be the Overview KPIs links and the Rules -> View alerts of this rule link.

### Parent issue
- https://github.com/wazuh/wazuh-dashboard-plugins/issues/6672

#### Rule alerts link
image

#### Overview KPIs links
image

### PR DESCRIPTION
**Pull request by Machi3mfl**
### Description

This PR adds a new behavior to create filters, add to the URL with the correct format and using it in a redirect link.

Closes #6739 

### Tasks

- [x] Create a method to make easier the generation of object filters

<details>

## Different filter operators based on the OSD search bar 

Screenshot 2024-06-05 at 13 14 48

### When the field is type number exists the IS BETWEEN and IS NOT BETWEEN operator

Screenshot 2024-06-06 at 11 18 00

```ts
// Examples

// "IS" Filter
const isFilter = PatternDataSourceFilterManager.createFilter(FILTER_OPERATOR.IS, 'agent.id','1','my-index');

// "IS NOT" Filter
const isNotFilter = PatternDataSourceFilterManager.createFilter(FILTER_OPERATOR.IS_NOT, ,'agent.id','1','my-index');

// "EXISTS" Filter
const existsFilter = PatternDataSourceFilterManager.createFilter(FILTER_OPERATOR.EXISTS, ,'agent.id','1','my-index');

// "DOES NOT EXIST" Filter
const doesNotExist = PatternDataSourceFilterManager.createFilter(FILTER_OPERATOR.DOESN_NOT_EXISTS,'agent.id','1','my-index');

// "IS ONE OF" Filter
const isOneOfFilter = PatternDataSourceFilterManager.createFilter(FILTER_OPERATOR.IS_ONE_OF','agent.id',['1'],'my-index');

// "IS NOT ONE OF" Filter
 const isNotOneOfFilter = PatternDataSourceFilterManager.createFilter(FILTER_OPERATOR.IS_NOT_ONE_OF,'agent.id',['1'],'my-index');

 const isBetweenFilter = PatternDataSourceFilterManager.createFilter(FILTER_OPERATOR.IS_BETWEEN,'agent.id',[1,2],'my-index');

 const isBetweenFilter = PatternDataSourceFilterManager.createFilter(FILTER_OPERATOR.IS_NOT_BETWEEN,'agent.id',[1,2],'my-index');
```
</details>

- [x] Create a method to transform a filter object to a URL filter format
- [x] Add unit tests

<details>

Screenshot 2024-06-06 at 12 00 24

</details>

- [x] Add method to create type range filter
- [x] Add rule stats links

# Tests

### Rule info redirect link to threat hunting with rule.id filter

<details>

https://github.com/wazuh/wazuh-dashboard-plugins/assets/6089438/ad2798c0-f410-43cd-b193-7990d8ac6d2d

</details>

### Overview `LAST 24 HOURS ALERTS` kpis redirect to discover with `rule.level is between` filter

<details>

https://github.com/wazuh/wazuh-dashboard-plugins/assets/6089438/ead9f5b1-1404-41ab-8abb-d3434a4c7ce6

</details>

### `View in rules` redirects in Flyout discover rule tab

<details>

https://github.com/wazuh/wazuh-dashboard-plugins/assets/6089438/c4886787-a02a-4a67-a17a-bc67c9b4f890

</details>

### Agent redirects in Techniques flyout discover table

<details>

https://github.com/wazuh/wazuh-dashboard-plugins/assets/6089438/a3e8a7cf-d0ae-4e7e-8fc2-f308c916d5a9

</details>

### Compliance visualization items redirect

<details>

https://github.com/wazuh/wazuh-dashboard-plugins/assets/6089438/18dec814-2cc3-4267-a151-b4c3639a709c

</details>

### Agent detail `SCA: Lastest scans` Policy table

<details>

### Not show the URL at the left bottom screen because is a on click event

https://github.com/wazuh/wazuh-dashboard-plugins/assets/6089438/23e68638-306f-4b9f-a0cd-f80cecca7105

</details>

### Check List
- [x] All tests pass
  - [x] `yarn test:jest`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: plugins/main/public/components/common/data-source/pattern/pattern-data-source-filter-manager.ts [hunk 1]
```diff
@@ -394,24 +395,29 @@ export class PatternDataSourceFilterManager
...
+    indexPatternId: string,
     controlledBy?: string,
-  ): tFilter {
-    switch (type) {
+  ) {
+    if (!key || !value || !indexPatternId) {
+      throw new Error('key, value and indexPatternId params are required');
+    }
+
+    if (type === 'is one of' || type === 'is not one of') {
```
Comment (asteriscos) [inline plugins/main/public/components/common/data-source/pattern/pattern-data-source-filter-manager.ts:409]: You are using strings as operators and it's used several times. I think you could improve this by defining an enum and exporting it, so the developers don't really need to know how the operator string is. E.g

```
enum FILTER_OPERATOR {
  IS = 'is',
  IS_NOT = 'is not',
  EXISTS = 'exists',
  DOES_NOT_EXISTS = 'does not exists',
  IS_ONE_OF = 'is one of',
  IS_NOT_ONE_OF = 'is not one of',
}
```

So if the string of the operator changes in the future we won't have to change all the implementations, except for the enum.

Comment (Machi3mfl) [reply plugins/main/public/components/common/data-source/pattern/pattern-data-source-filter-manager.ts]: Changed in https://github.com/wazuh/wazuh-dashboard-plugins/pull/6745/commits/bbfb7a2a794acf22a1682d447a2c0494a6b836e4

File: plugins/main/public/components/common/data-source/pattern/pattern-data-source-filter-manager.ts [hunk 2]
```diff
@@ -514,27 +517,42 @@ export class PatternDataSourceFilterManager
...
+      type,
+      key,
+      value,
+      this.dataSource.id,
+      controlledBy,
+    );
+  }
+
   /**
    * Return a simple filter object with the field, value and index pattern received
```
Comment (JuanGarriuz) [inline plugins/main/public/components/common/data-source/pattern/pattern-data-source-filter-manager.ts:546]: The name of the received parameters should be changed from `indexPatternTitle` to `indexPatternId`.

File: plugins/main/public/components/common/data-source/pattern/pattern-data-source-filter-manager.ts [hunk 3]
```diff
@@ -514,27 +517,42 @@ export class PatternDataSourceFilterManager
...
-          undefined,
-          controlledBy,
-        );
+        throw new Error('Invalid filter type');
     }
   }
 
+  /**
+   * Returns a filter with the field and value received
+   * @param field
```
Comment (JuanGarriuz) [inline plugins/main/public/components/common/data-source/pattern/pattern-data-source-filter-manager.ts:526]: `Field` param is not used

**Comment by asteriscos** [comment]
## Test :green_circle: 
### Rule info redirect link to threat hunting with rule.id filter

<details>

Peek 2024-06-07 17-13


</details>

### Overview `LAST 24 HOURS ALERTS` kpis redirect to discover with `rule.level is between` filter

<details>

Peek 2024-06-07 17-13


</details>

### `View in rules` redirects in Flyout discover rule tab

<details>

Peek 2024-06-07 17-22


</details>

### Agent redirects in Techniques flyout discover table

<details>

Peek 2024-06-07 17-28



</details>

### Compliance visualization items redirect

<details>


Peek 2024-06-07 17-38



</details>

### Agent detail `SCA: Lastest scans` Policy table

<details>

### Not show the URL at the left bottom screen because is a on click event

Peek 2024-06-07 17-36

</details>

**Pull request merged**